### IMPORTING LIBRARIES

In [44]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import statsmodels.api as sm

### READING THE DATASET

In [45]:
df = pd.read_csv('C:\\Users\\user\\Downloads\\US_Heart_Patients.csv')
df.head()

male  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39        4.0              0         0.0     0.0                0   
1     0   46        2.0              0         0.0     0.0                0   
2     1   48        1.0              1        20.0     0.0                0   
3     0   61        3.0              1        30.0     0.0                0   
4     0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

In [46]:
df.shape

(4240, 16)

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4240 entries, 0 to 4239
Data columns (total 16 columns):
male               4240 non-null int64
age                4240 non-null int64
education          4135 non-null float64
currentSmoker      4240 non-null int64
cigsPerDay         4211 non-null float64
BPMeds             4187 non-null float64
prevalentStroke    4240 non-null int64
prevalentHyp       4240 non-null int64
diabetes           4240 non-null int64
totChol            4190 non-null float64
sysBP              4240 non-null float64
diaBP              4240 non-null float64
BMI                4221 non-null float64
heartRate          4239 non-null float64
glucose            3852 non-null float64
TenYearCHD         4240 non-null int64
dtypes: float64(9), int64(7)
memory usage: 530.1 KB


### CHECKING FOR NULL VALUES

In [48]:
df.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

###### The events happening at 6 sigma and above are called Black Swan Events in Statistics.

### DROPPING NULL VALUES

In [49]:
df2 = df.dropna()

In [50]:
df2.shape

(3658, 16)

### CORRELATION

In [51]:
df2.corr()

male       age  education  currentSmoker  cigsPerDay  \
male             1.000000 -0.024412   0.017729       0.206114    0.330322   
age             -0.024412  1.000000  -0.159499      -0.210771   -0.189295   
education        0.017729 -0.159499   1.000000       0.025251    0.013790   
currentSmoker    0.206114 -0.210771   0.025251       1.000000    0.773913   
cigsPerDay       0.330322 -0.189295   0.013790       0.773913    1.000000   
BPMeds          -0.052124  0.134732  -0.013739      -0.051923   -0.046504   
prevalentStroke -0.002312  0.050893  -0.030386      -0.038150   -0.036286   
prevalentHyp     0.001424  0.306239  -0.078828      -0.108078   -0.070460   
diabetes         0.013819  0.109092  -0.039627      -0.041849   -0.036961   
totChol         -0.070321  0.268252  -0.013522      -0.051034   -0.030400   
sysBP           -0.045358  0.388267  -0.124360      -0.134428   -0.094781   
diaBP            0.051872  0.208283  -0.058059      -0.115955   -0.056746   
BMI              0.073111  0.137511  -0.137555      -0.159821   -0.087395   
heartRate       -0.115285 -0.002722  -0.064214       0.050841    0.064030   
glucose          0.002940  0.118349  -0.031998      -0.053242   -0.053726   
TenYearCHD       0.091688  0.233983  -0.063279       0.019165    0.052014   

                   BPMeds  prevalentStroke  prevalentHyp  diabetes   totChol  \
male            -0.052124        -0.002312      0.001424  0.013819 -0.070321   
age              0.134732         0.050893      0.306239  0.109092  0.268252   
education       -0.013739        -0.030386     -0.078828 -0.039627 -0.013522   
currentSmoker   -0.051923        -0.038150     -0.108078 -0.041849 -0.051034   
cigsPerDay      -0.046504        -0.036286     -0.070460 -0.036961 -0.030400   
BPMeds           1.000000         0.113125      0.262910  0.049066  0.094083   
prevalentStroke  0.113125         1.000000      0.066057  0.009625  0.012736   
prevalentHyp     0.262910         0.066057      1.000000  0.080556  0.166655   
diabetes         0.049066         0.009625      0.080556  1.000000  0.048451   
totChol          0.094083         0.012736      0.166655  0.048451  1.000000   
sysBP            0.271263         0.061070      0.697675  0.102552  0.219925   
diaBP            0.199630         0.055834      0.617734  0.050686  0.174422   
BMI              0.105642         0.036496      0.302949  0.089009  0.121056   
heartRate        0.012889        -0.017019      0.146818  0.060984  0.093053   
glucose          0.054232         0.016061      0.086942  0.614820  0.049884   
TenYearCHD       0.089152         0.048366      0.181387  0.093431  0.091338   

                    sysBP     diaBP       BMI  heartRate   glucose  TenYearCHD  
male            -0.045358  0.051872  0.073111  -0.115285  0.002940    0.091688  
age              0.388267  0.208283  0.137511  -0.002722  0.118349    0.233983  
education       -0.124360 -0.058059 -0.137555  -0.064214 -0.031998   -0.063279  
currentSmoker   -0.134428 -0.115955 -0.159821   0.050841 -0.053242    0.019165  
cigsPerDay      -0.094781 -0.056746 -0.087395   0.064030 -0.053726    0.052014  
BPMeds           0.271263  0.199630  0.105642   0.012889  0.054232    0.089152  
prevalentStroke  0.061070  0.055834  0.036496  -0.017019  0.016061    0.048366  
prevalentHyp     0.697675  0.617734  0.302949   0.146818  0.086942    0.181387  
diabetes         0.102552  0.050686  0.089009   0.060984  0.614820    0.093431  
totChol          0.219925  0.174422  0.121056   0.093053  0.049884    0.091338  
sysBP            1.000000  0.786669  0.330917   0.184797  0.134651    0.222821  
diaBP            0.786669  1.000000  0.385348   0.178744  0.063540    0.150124  
BMI              0.330917  0.385348  1.000000   0.074131  0.083683    0.082055  
heartRate        0.184797  0.178744  0.074131   1.000000  0.097074    0.020514  
glucose          0.134651  0.063540  0.083683   0.097074  1.000000    0.121990  
TenYearCHD       0.222821  0.150124  0.082055   0.0205

###### RENAMING MALE COLUMN TO GENDER

In [52]:
df2.rename(columns={'male':'gender'},inplace=True)

C:\Users\user\Anaconda3\lib\site-packages\pandas\core\frame.py:4025: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


In [53]:
df2.corr()

gender       age  education  currentSmoker  cigsPerDay  \
gender           1.000000 -0.024412   0.017729       0.206114    0.330322   
age             -0.024412  1.000000  -0.159499      -0.210771   -0.189295   
education        0.017729 -0.159499   1.000000       0.025251    0.013790   
currentSmoker    0.206114 -0.210771   0.025251       1.000000    0.773913   
cigsPerDay       0.330322 -0.189295   0.013790       0.773913    1.000000   
BPMeds          -0.052124  0.134732  -0.013739      -0.051923   -0.046504   
prevalentStroke -0.002312  0.050893  -0.030386      -0.038150   -0.036286   
prevalentHyp     0.001424  0.306239  -0.078828      -0.108078   -0.070460   
diabetes         0.013819  0.109092  -0.039627      -0.041849   -0.036961   
totChol         -0.070321  0.268252  -0.013522      -0.051034   -0.030400   
sysBP           -0.045358  0.388267  -0.124360      -0.134428   -0.094781   
diaBP            0.051872  0.208283  -0.058059      -0.115955   -0.056746   
BMI              0.073111  0.137511  -0.137555      -0.159821   -0.087395   
heartRate       -0.115285 -0.002722  -0.064214       0.050841    0.064030   
glucose          0.002940  0.118349  -0.031998      -0.053242   -0.053726   
TenYearCHD       0.091688  0.233983  -0.063279       0.019165    0.052014   

                   BPMeds  prevalentStroke  prevalentHyp  diabetes   totChol  \
gender          -0.052124        -0.002312      0.001424  0.013819 -0.070321   
age              0.134732         0.050893      0.306239  0.109092  0.268252   
education       -0.013739        -0.030386     -0.078828 -0.039627 -0.013522   
currentSmoker   -0.051923        -0.038150     -0.108078 -0.041849 -0.051034   
cigsPerDay      -0.046504        -0.036286     -0.070460 -0.036961 -0.030400   
BPMeds           1.000000         0.113125      0.262910  0.049066  0.094083   
prevalentStroke  0.113125         1.000000      0.066057  0.009625  0.012736   
prevalentHyp     0.262910         0.066057      1.000000  0.080556  0.166655   
diabetes         0.049066         0.009625      0.080556  1.000000  0.048451   
totChol          0.094083         0.012736      0.166655  0.048451  1.000000   
sysBP            0.271263         0.061070      0.697675  0.102552  0.219925   
diaBP            0.199630         0.055834      0.617734  0.050686  0.174422   
BMI              0.105642         0.036496      0.302949  0.089009  0.121056   
heartRate        0.012889        -0.017019      0.146818  0.060984  0.093053   
glucose          0.054232         0.016061      0.086942  0.614820  0.049884   
TenYearCHD       0.089152         0.048366      0.181387  0.093431  0.091338   

                    sysBP     diaBP       BMI  heartRate   glucose  TenYearCHD  
gender          -0.045358  0.051872  0.073111  -0.115285  0.002940    0.091688  
age              0.388267  0.208283  0.137511  -0.002722  0.118349    0.233983  
education       -0.124360 -0.058059 -0.137555  -0.064214 -0.031998   -0.063279  
currentSmoker   -0.134428 -0.115955 -0.159821   0.050841 -0.053242    0.019165  
cigsPerDay      -0.094781 -0.056746 -0.087395   0.064030 -0.053726    0.052014  
BPMeds           0.271263  0.199630  0.105642   0.012889  0.054232    0.089152  
prevalentStroke  0.061070  0.055834  0.036496  -0.017019  0.016061    0.048366  
prevalentHyp     0.697675  0.617734  0.302949   0.146818  0.086942    0.181387  
diabetes         0.102552  0.050686  0.089009   0.060984  0.614820    0.093431  
totChol          0.219925  0.174422  0.121056   0.093053  0.049884    0.091338  
sysBP            1.000000  0.786669  0.330917   0.184797  0.134651    0.222821  
diaBP            0.786669  1.000000  0.385348   0.178744  0.063540    0.150124  
BMI              0.330917  0.385348  1.000000   0.074131  0.083683    0.082055  
heartRate        0.184797  0.178744  0.074131   1.000000  0.097074    0.020514  
glucose          0.134651  0.063540  0.083683   0.097074  1.000000    0.121990  
TenYearCHD       0.222821  0.150124  0.082055   0.02

### ADDING A CONSTANT

In [55]:
from statsmodels.tools import add_constant
df_constant = add_constant(df2)
df_constant.head()

C:\Users\user\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


const  gender  age  education  currentSmoker  cigsPerDay  BPMeds  \
0    1.0       1   39        4.0              0         0.0     0.0   
1    1.0       0   46        2.0              0         0.0     0.0   
2    1.0       1   48        1.0              1        20.0     0.0   
3    1.0       0   61        3.0              1        30.0     0.0   
4    1.0       0   46        3.0              1        23.0     0.0   

   prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0                0             0         0    195.0  106.0   70.0  26.97   
1                0             0         0    250.0  121.0   81.0  28.73   
2                0             0         0    245.0  127.5   80.0  25.34   
3                0             1         0    225.0  150.0   95.0  28.58   
4                0             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose  TenYearCHD  
0       80.0     77.0           0  
1       95.0     76.0           0  
2       75.0     70.0           0  
3       65.0    103.0           1  
4       85.0     85.0           0

### SPLITTING x AND y

In [14]:
x = df.drop('TenYearCHD',axis=1)
y = df['TenYearCHD']

### MODELLING

In [20]:
x = df_constant.drop('TenYearCHD',axis=1)
y = df_constant['TenYearCHD']

### LOGISTIC REGRESSION

In [57]:
cols = df_constant.columns[:-1]
model = sm.Logit(df2.TenYearCHD,df_constant[cols])
result = model.fit()
result.summary()

Optimization terminated successfully.
         Current function value: 0.376500
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3658
Model:                          Logit   Df Residuals:                     3642
Method:                           MLE   Df Model:                           15
Date:                Sun, 19 Apr 2020   Pseudo R-squ.:                  0.1175
Time:                        19:37:33   Log-Likelihood:                -1377.2
converged:                       True   LL-Null:                       -1560.6
                                        LLR p-value:                 6.676e-69
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              -8.3282      0.715    -11.640      0.000      -9.730      -6.926
gender              0.5553      0.109      5.093      0.000       0.342       0.769
age                 0.0635      0.007      9.509      0.000       0.050       0.077
education          -0.0478      0.049     -0.967      0.334      -0.145       0.049
currentSmoker       0.0716      0.157      0.457      0.648      -0.236       0.379
cigsPerDay          0.0179      0.006      2.872      0.004       0.006       0.030
BPMeds              0.1625      0.234      0.693      0.488      -0.297       0.622
prevalentStroke     0.6937      0.490      1.417      0.157      -0.266       1.653
prevalentHyp        0.2342      0.138      1.697      0.090      -0.036       0.505
diabetes            0.0392      0.316      0.124      0.901      -0.579       0.658
totChol             0.0023      0.001      2.070      0.038       0.000       0.005
sysBP               0.0154      0.004      4.044      0.000       0.008       0.023
diaBP              -0.0042      0.006     -0.646      0.518      -0.017       0.008
BMI                 0.0067      0.013      0.523      0.601      -0.018       0.032
heartRate          -0.0032      0.004     -0.771      0.441      -0.012       0.005
glucose             0.0071      0.002      3.190      0.001       0.003       0.012
===================================================================================
"""

### BACKWARD ELIMINATION FOR LOGISTIC REGRESSION

In [59]:
def back_feature_elim(data_frame,dep_var,col_list):
    
    while len(col_list)>0:
        model = sm.Logit(dep_var,data_frame[col_list])
        result = model.fit(disp=0)
        largest_pvalue = round(result.pvalues,3).nlargest(1)
        if largest_pvalue[0]<(0.05):
            return result
            break
        else:
            col_list = col_list.drop(largest_pvalue.index)
result = back_feature_elim(df_constant,df2.TenYearCHD,cols)     

In [61]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3658
Model:                          Logit   Df Residuals:                     3651
Method:                           MLE   Df Model:                            6
Date:                Sun, 19 Apr 2020   Pseudo R-squ.:                  0.1148
Time:                        19:39:34   Log-Likelihood:                -1381.4
converged:                       True   LL-Null:                       -1560.6
                                        LLR p-value:                 2.408e-74
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -9.1353      0.475    -19.213      0.000     -10.067      -8.203
gender         0.5617      0.107      5.258      0.000       0.352       0.771
age            0.0660      0.006     10.267      0.000       0.053       0.079
cigsPerDay     0.0192      0.004      4.606      0.000       0.011       0.027
totChol        0.0023      0.001      2.031      0.042    8.01e-05       0.004
sysBP          0.0175      0.002      8.155      0.000       0.013       0.022
glucose        0.0073      0.002      4.343      0.000       0.004       0.011
==============================================================================
"""

In [72]:
x_final = df2[['gender','age','cigsPerDay','totChol','sysBP','glucose']]
y = df2['TenYearCHD']

### TRAIN TEST SPLIT

In [83]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x_final,y,test_size = 0.3,random_state = 0)

In [84]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(x_train,y_train)
y_pred = model.predict(x_test)

In [82]:
x_train.shape,y_train.shape

((2560, 6), (1098, 6))

### EVALUATION METRICS

In [85]:
from sklearn import metrics

In [87]:
accuracy = metrics.accuracy_score(y_test,y_pred)
accuracy

0.8406193078324226

In [88]:
cm = metrics.confusion_matrix(y_test,y_pred)
cm

array([[908,   8],
       [167,  15]], dtype=int64)

#### SENSITIVITY(TRUE POSITIVE RATE)

In [91]:
tpr = cm[1,1]/cm[1,:].sum()
tpr

0.08241758241758242

#### SPECIFICITY(TRUE NEGATIVE RATE)

In [93]:
tnr = cm[0,0]/cm[0,:].sum()
tnr

0.9912663755458515

###### NOTE: The coefficient for totChol is very less(0.0023). So we will remove it and model again.

In [98]:
x_final = df2[['gender','age','cigsPerDay','sysBP','glucose']]
y = df2['TenYearCHD']

In [99]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x_final,y,test_size = 0.3,random_state = 0)

In [100]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(x_train,y_train)
y_pred = model.predict(x_test)

In [101]:
accuracy = metrics.accuracy_score(y_test,y_pred)
accuracy

0.8415300546448088

In [102]:
cm = metrics.confusion_matrix(y_test,y_pred)
cm

array([[907,   9],
       [165,  17]], dtype=int64)

In [105]:
tpr = cm[1,1]/cm[1,:].sum()
tpr*100

9.340659340659341

In [106]:
tnr = cm[0,0]/cm[0,:].sum()
tnr*100

99.0174672489083

###### As we can see, there are no major changes after dropping totChol. Its contribution is very less. So, we drop it.

In [132]:
np.exp(result.params)

const         0.000108
gender        1.753586
age           1.068190
cigsPerDay    1.019413
totChol       1.002283
sysBP         1.017682
glucose       1.007310
dtype: float64

###### KEYNOTE:
###### 1)To make inferences from the Logit summary, always take exponential of the coefficients. The exponential is taken because 
###### ( P / 1 - P ) = e * * (B0+B1*x1+B2*x2) . 
###### 2)The risk decreases when the exponential value is less than 1. If the exponential value is above 1, then the risk increases.

# INFERENCE:

###### 1)GENDER:
###### If the patient is a MALE, the risk of getting a CHD increases by 75 %. Because 1 is treated as male in the gender column.

###### 2)AGE:
###### For every one unit increase in AGE, the risk of getting a CHD increases by 6.8 %

###### 3)CIGERATTES PER DAY:
###### For every one CIGERATTES PER DAY, the risk of getting a CHD increases by 1.9 %  

###### 4)TOTCHOL:
###### For every one unit increase in TOTCHOL, the risk of getting a CHD increases by 0.2 %

###### 4)SYSTOLIC BP:
###### For every one unit increase in SYSTOLIC BP, the risk of getting a CHD increases by 1.7 %

###### 5)GLUCOSE:
###### For every one unit increase in GLUCOSE, the risk of getting a CHD increases by 0.7 %